Cleanup and refactor FnCtxt::report_no_match_method_error

rperier opened the pull request:
As discussed on zulip with @lcnr, this is a proposal for refactorizing this method.


See [#t-compiler/help > (PR #144674) Merge multiple suggestions into a single @ 💬](https://rust-lang.zulipchat.com/#narrow/channel/182449-t-compiler.2Fhelp/topic/.28PR.20.23144674.29.20Merge.20multiple.20suggestions.20into.20a.20single/near/539991695)

<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r? <reviewer name>
-->
<!-- homu-ignore:end -->

rperier commented:
r? @lcnr

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/19172914254/job/54810092617) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/54810092617) [(plain)](https://github.com/rust-lang/rust/commit/177bb33825b2d66fb5e5c0e44c2727cdfbb10f3a/checks/54810092617/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
 Documenting proc_macro_test v0.1.0 (/checkout/tests/rustdoc-gui/src/proc_macro_test)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.80s
   Generated /checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc-gui/doc/proc_macro_test/index.html
npm WARN deprecated puppeteer@22.15.0: < 24.10.2 is no longer supported
npm ERR! code 127
npm ERR! git dep preparation failed
npm ERR! command /node/bin/node /node/lib/node_modules/npm/bin/npm-cli.js install --force --cache=/home/user/.npm --prefer-offline=false --prefer-online=false --offline=false --no-progress --no-save --no-audit --include=dev --include=peer --include=optional --no-package-lock-only --no-dry-run
npm ERR! npm WARN using --force Recommended protections disabled.
npm ERR! npm WARN deprecated inflight@1.0.6: This module is not supported, and leaks memory. Do not use it. Check out lru-cache if you want a good and tested way to coalesce async requests by a key value, which is much more comprehensive and powerful.
npm ERR! npm WARN deprecated glob@7.2.3: Glob versions prior to v9 are no longer supported
npm ERR! npm WARN deprecated glob@7.2.3: Glob versions prior to v9 are no longer supported
npm ERR! npm WARN deprecated readdir-scoped-modules@1.1.0: This functionality has been moved to @npmcli/fs
npm ERR! npm WARN deprecated read-package-json@2.1.2: This package is no longer supported. Please use @npmcli/package-json instead.
npm ERR! npm WARN deprecated debuglog@1.0.1: Package no longer supported. Contact Support at https://www.npmjs.com/support for more info.
npm ERR! npm WARN deprecated read-installed@4.0.3: This package is no longer supported.
npm ERR! npm ERR! code 127
npm ERR! npm ERR! path /home/user/.npm/_cacache/tmp/git-cloneXXXXXX54ZpIE/node_modules/rollup
npm ERR! npm ERR! command failed
npm ERR! npm ERR! command sh -c patch-package
npm ERR! npm ERR! sh: 1: patch-package: not found
npm ERR! 
npm ERR! npm ERR! A complete log of this run can be found in: /home/user/.npm/_logs/2025-11-07T16_34_13_945Z-debug-0.log

npm ERR! A complete log of this run can be found in: /home/user/.npm/_logs/2025-11-07T16_34_06_629Z-debug-0.log
npm install did not exit successfully

thread 'main' (59440) panicked at src/tools/rustdoc-gui-test/src/main.rs:63:10:
unable to install browser-ui-test: Custom { kind: Other, error: "npm install returned exit code exit status: 127" }
stack backtrace:
   0: __rustc::rust_begin_unwind
             at /rustc/3b4dd9bf1410f8da6329baa36ce5e37673cbbd1f/library/std/src/panicking.rs:698:5
   1: core::panicking::panic_fmt
             at /rustc/3b4dd9bf1410f8da6329baa36ce5e37673cbbd1f/library/core/src/panicking.rs:80:14
```

</details>

lcnr commented on compiler/rustc_hir_typeck/src/method/suggest.rs:
```diff
@@ -936,21 +876,53 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+        item_kind: &str,
+        mode: Mode,
+        source: SelfSource<'tcx>,
+        unsatisfied_predicates: &Vec<(
```
`&Vec` should pretty much always be a slice?

lcnr commented on compiler/rustc_hir_typeck/src/method/suggest.rs:
```diff
@@ -1077,6 +1049,211 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+        ) {
+            return guar;
+        }
+        span = item_ident.span;
```
that's weird. why? and could we just use `item_ident.span` directly below this point?

rperier commented on compiler/rustc_hir_typeck/src/method/suggest.rs:
```diff
@@ -1077,6 +1049,211 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+        };
+        if rcvr_ty.references_error() {
+            err.downgrade_to_delayed_bug();
+        }
```
Because you need `err` , and the error is created just  above

rperier committed:
Add sub-fn for method call annotation in FnCtxt::report_no_match_method_error

Currently this method is quiet long and complex, this commit
refactors it and improves its readability by adding sub-fn

rperier committed:
Add sub-fn for static method candidates in FnCtxt::report_no_match_method_error

Currently this method is quiet long and complex, this commit
refactors it and improves its readability by adding sub-fn

rperier committed:
Add sub-fn for unsatisfied ty or trait in FnCtxt::report_no_match_method_error

Currently this method is quiet long and complex, this commit
refactors it and improves its readability by adding sub-fn

rperier committed:
Add sub-fn for surround method calls in FnCtxt::report_no_match_method_error

Currently this method is quiet long and complex, this commit
refactors it and improves its readability by adding sub-fn

rperier committed:
Add sub-fn to find possible candidates for method in FnCtxt::report_no_match_method_error

Currently this method is quiet long and complex, this commit
refactors it and improves its readability by adding sub-fn

rperier committed:
Add sub-fn for confusable or similarly named methods in FnCtxt::report_no_match_method_error

Currently this method is quiet long and complex, this commit
refactors it and improves its readability by adding sub-fn

rperier committed:
Add sub-fn for not found methods with unsatisfied bounds in FnCtxt::report_no_match_method_error

Currently this method is quiet long and complex, this commit
refactors it and improves its readability by adding sub-fn

rperier committed:
Refactor and cleanup FnCtxt::report_no_match_method_error

Currently this method is quiet long and complex, this commit
improves its readability, refactor and cleanup few things

rperier commented:
I have splitted the changes into severals commits, it will be easier to review. I have included most of the requested changes already.
